spec(consensus): update Block Time documentation in spec (#2316)
cometbft/cometbft: spec/consensus/time.md

---
order: 2
---
# Time

CometBFT provides a Byzantine fault-tolerant source of time.

Time in CometBFT is defined with the [`Time`][spec-time] field of the
block [`Header`][spec-header].

## Properties

The Time produced by CometBFT satisfies the following properties:

- **Time Monotonicity**: time is monotonically increasing.  More precisely, given
  two block headers `H1` of height `h1` and `H2` of height `h2`,
  it is guaranteed that if `h2 > h1` then `H2.Time > H1.Time`.

- **Byzantine Fault Tolerance**: faulty or malicious processes should not be able
  to arbitrarily influence (increase or decrease) the block Time.

>   In other words, the Time of blocks should be defined by correct processes.

Reviewer: I have a (small) problem with this sentence. If a byzantine process (which is byzantine for the whole run) decides to use a "correct" time value, this sentence is not correct, is it?
I know it can't happen in BFT-Time, but it _could_ in PBTS if the byzantine node in question is the proposer.

Author: In this case they are not arbitrarily influencing the block time, are them?

Reviewer: Agreed, that's about the first sentence, which I'm fine with.
The problem is the second sentence, which says "the Time of blocks should be defined by correct processes".

Reviewer: As is starts with "In other words...", consider removing it

In addition, the Time produced by CometBFT is expected, by external observers, to provide:

- **Relation to real time**: block times bear some resemblance to real time.
  In other words, block times should represent, within some reasonable accuracy,

>   the actual time at which blocks were produced.

Reviewer 2: This definition is problematic but I have no better one to offer.

Author: Indeed...

Reviewer: An idea would be to express it with a bound, unknown but existing. Something like:

_ block times bear some resemblance to real time. For every decided block `H` of height `h`, there exists an unknown bound `B` such that, `| H.Timestamp - real_time | < B`, where `real_time` is the actual time when `H` was first proposed.

Anyway, I'm not sure about this. You may well decide to keep the current text, I'm OK with that.

Author: We had an `ACCURACY` parameter in the original spec. But relating it to `PRECISION` is not really trivial, although one can assume that `ACCURACY <= 2*PRECISION`.

In this case, we have to add a `2` factor for all `PRECISION` entries in the timely properties. Which has rendered the spec more complex, and we are not sure if this reasoning is right.

Author: See: https://github.com/cometbft/cometbft/blob/master/spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md#accuracy

Reviewer: My suggestion is really basic... definitely not proposing to change the sysmodel in the spec. It was just another possible wording for the "Relation to real time" definition.

## Implementations

CometBFT implements two algorithms for computing block times:

- [BFT Time][bft-time]: the algorithm adopted in versions up to `v0.38.x`;

>   available, in legacy mode, in version `v1.x`.

Reviewer: ```suggestion
  available, in legacy mode, in version `v1.x` and later.
```

Author: Are we making this promise?

Reviewer: Are we gonna be able to nuke it, and still support blocksyncing from arbitrarily in the past? I don't think so.
Anyway, I'm also OK without the "and later".

> - [Proposer-Based Timestamps][pbts-spec]: introduced in version `v1.x`,
>   intended to be a replacement for BFT Time.

Reviewer 2: ```suggestion
- [Proposer-Based Timestamps][pbts-spec]: introduced in version `v1.x`,
  as a replacement for BFT Time.
```
